From abdc4ff654ca6f4a36650826be43f3a43ab56cf2 Mon Sep 17 00:00:00 2001
From: devmfc <devmfc@mvand.net>
Date: Wed, 22 Jan 2025 21:00:25 +0100
Subject: [PATCH 32/32] [DEVMFC] drm/meson: fix component unbind in error path

Only call component_unbind_all if component_bind_all has actually
succeeded.

This prevents a crash when an error occurs before calling
component_bind_all. This happens for example when meson_encoder_cvbs_probe
returns EPROBE_DEFER.
---
 drivers/gpu/drm/meson/meson_drv.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/drivers/gpu/drm/meson/meson_drv.c b/drivers/gpu/drm/meson/meson_drv.c
index 8f378efde3..2e427d05b9 100644
--- a/drivers/gpu/drm/meson/meson_drv.c
+++ b/drivers/gpu/drm/meson/meson_drv.c
@@ -213,7 +213,9 @@ static int meson_drv_bind_master(struct device *dev, bool has_components)
 	const struct meson_drm_match_data *match;
 	struct meson_drm *priv;
 	struct drm_device *drm;
+	bool do_unbind = false;
 	struct resource *res;
+	
 	void __iomem *regs;
 	int ret, i;
 
@@ -409,7 +411,6 @@ static int meson_drv_bind_master(struct device *dev, bool has_components)
 	}
 
 	/* Encoder Initialization */
-
 	ret = meson_encoder_cvbs_probe(priv);
 	if (ret)
 		goto exit_afbcd;
@@ -419,9 +420,9 @@ static int meson_drv_bind_master(struct device *dev, bool has_components)
 		if (ret) {
 			dev_err(drm->dev, "Couldn't bind all components\n");
 			/* Do not try to unbind */
-			has_components = false;
 			goto exit_afbcd;
 		}
+		do_unbind = true;
 	}
 
 	ret = meson_encoder_hdmi_probe(priv);
@@ -484,7 +485,7 @@ static int meson_drv_bind_master(struct device *dev, bool has_components)
 	meson_encoder_hdmi_remove(priv);
 	meson_encoder_cvbs_remove(priv);
 
-	if (has_components)
+	if (do_unbind)
 		component_unbind_all(dev, drm);
 
 	return ret;
-- 
2.46.0

